Add Accounting Admin and Enterprise Admin UIs for managing SSO

biyeun opened the pull request:
## Summary
Lots of commits (!!!), but all very small and well-commented.

This PR introduces...

Accounting Admin UIs for:
- looking up `IdentityProvider`s
- creating an `IdentityProvider`
- editing an `IdentityProvider`, along with specifying `AuthenticatedEmailDomain`(s) and `UserExemptFromSingleSignOn`(s)

Enterprise Admin UIs for:
- editing an `IdentityProvider` (if `is_editable` is set to `True`)
- managing the list of `UserExemptFromSingleSignOn`

It also introduces a new enterprise app decorator that adds the account to the request object. In a future PR, I would like to decorate the other enterprise views with this and add the authentication check from `get_account_or_404` here rather than keeping it in the get method.

There's a lot of complexity in the user-interaction of the UIs, so I feel that just adding a bunch of screenshots here in the description will lose quite a bit of context. The screenshots are here as a summary, but I encourage anyone to test this out on staging with the following steps:

1) Add yourself to the [`ENTERPRISE_SSO` feature flag](https://staging.commcarehq.org/hq/flags/edit/enterprise_sso/)
2) [List of Identity Providers](https://staging.commcarehq.org/hq/accounting/identity_providers/) in Accounting Admin (requires accounting admin permission on staging)
3) [Edit UI of existing Identity Provider](https://staging.commcarehq.org/hq/accounting/identity_provider/1/) (requires Accounting Admin permission)
4) [Edit UI of Enterprise Admin](https://staging.commcarehq.org/a/vault-wax/enterprise/sso/)

## Feature Flag
`ENTERPRISE_SSO`

## Product Description
![Screen Shot 2021-02-16 at 6 54 16 PM](https://user-images.githubusercontent.com/716573/108136627-03fd7880-7080-11eb-984c-d9c12c2b3b89.png)
![Screen Shot 2021-02-16 at 6 54 40 PM](https://user-images.githubusercontent.com/716573/108136673-111a6780-7080-11eb-9fa7-2f83df47ad57.png)
![Screen Shot 2021-02-16 at 6 55 12 PM](https://user-images.githubusercontent.com/716573/108136700-255e6480-7080-11eb-92ab-f75596575821.png)
![Screen Shot 2021-02-16 at 6 55 44 PM](https://user-images.githubusercontent.com/716573/108136860-6ce4f080-7080-11eb-8afd-7934a9bf3bd6.png)
![Screen Shot 2021-02-16 at 6 56 32 PM](https://user-images.githubusercontent.com/716573/108136862-6d7d8700-7080-11eb-9cf0-ad1a325b062a.png)
![Screen Shot 2021-02-16 at 6 56 20 PM](https://user-images.githubusercontent.com/716573/108136863-6d7d8700-7080-11eb-81ff-febdbd68c1da.png)
![Screen Shot 2021-02-16 at 6 55 50 PM](https://user-images.githubusercontent.com/716573/108136868-6eaeb400-7080-11eb-8b62-5a890a830d19.png)
![Screen Shot 2021-02-16 at 7 01 46 PM](https://user-images.githubusercontent.com/716573/108137221-16c47d00-7081-11eb-8e07-c2dd8616e00a.png)
![Screen Shot 2021-02-16 at 7 01 57 PM](https://user-images.githubusercontent.com/716573/108137226-175d1380-7081-11eb-9ea9-072ec6149a7e.png)

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage
Added new tests for the forms, async handlers required for managing `AuthenticatedEmailDomain` and `UserExemptFromSingleSignOn`, which I think are the most critical parts to making sure the `IdentityProviders` are valid before being set to active. Also added test for new decorator in enterprise app.

### QA Plan
No QA needed now, as this is all behind a feature flag

### Safety story
Behind a feature flag and there are tons of tests. Does not affect core workflows.

### Rollback instructions
- [x] This PR can be reverted after deploy with no further considerations

biyeun committed:
add an accounting interface for managing identity providers

biyeun committed:
allow a custom async handler to be specified in accounting widgets js

biyeun committed:
add async handler for upcoming identity provider view

biyeun committed:
add view for creating identity providers in the accounting admin

biyeun committed:
ad utilities for quickly fetching full SAML2 endpoint URLs

biyeun committed:
fix error

biyeun committed:
add async handlers for managing sso exempt users and email domains and corresponding js/knockout models

biyeun committed:
refrence new SAML2 url utils in config

biyeun committed:
add edit form for identity provider in accounting admin

biyeun committed:
add edit identity provider view in accounting admin

biyeun committed:
implement certificate download response

biyeun committed:
fix some minor styling quirks

biyeun committed:
add permissions check decorator for enterprise views
todo: apply to other enterprise views in a followup PR

biyeun committed:
add base enterprise class-based view

biyeun committed:
add manage sso enterprise admin view

biyeun committed:
make part of form with service provider details reusable. also add some fancier styling

biyeun committed:
not needed

biyeun committed:
small copy changes

biyeun committed:
add edit identity provider view to the enterprise dashboard

biyeun committed:
return certificate responses

biyeun committed:
use datetimepicker widget

biyeun committed:
prevent sso exempt users from being less than one if is_editable is active

biyeun committed:
add property for sso users to identity provider

biyeun committed:
add active and editable checks in forms

biyeun committed:
make sure view uses datetimepicker

biyeun committed:
remove prepend text on email field

biyeun committed:
add convenience link to enterprise admin view

biyeun committed:
text and placeholder updates

biyeun committed:
change to charfield due to datetime string

biyeun committed:
forgotten import whoops

biyeun committed:
set initial values

biyeun committed:
actually update identity provider on form submission

biyeun committed:
validate with regex. should probably make a better global util for this

biyeun committed:
test new enterprise decorator

biyeun committed:
actually raise validation error rather than string

biyeun committed:
i can spell, i sware...er, swear 👀

biyeun committed:
whoops actually update the name in the idp from the form

biyeun committed:
return values in alphebetical order

biyeun committed:
fix error when a linked_subscription isn't found

biyeun committed:
make TIME_FORMAT constant for further use in tests

biyeun committed:
add tests for SSO forms

biyeun committed:
add tests for sso async handlers

biyeun committed:
lint

biyeun committed:
lint

biyeun commented:
I disagree with the automatic risk labels being set as all new code is behind a feature flag!

biyeun commented:
ah, I think the high-risk tag is coming from the `@use_datetimepicker` decorator I've added. arguably that should be its own PR. do people feel strongly about separating this out? those commits are easy to cherry-pick into their own thing... there are a few other additions which I'm happy to split up. didn't realize just how massive this work was until pushing this upstream 👀

biyeun commented:
split out the datetimepicker stuff here: https://github.com/dimagi/commcare-hq/pull/29187

orangejenny commented on corehq/apps/hqwebapp/static/hqwebapp/js/requirejs_config.js:
```diff
@@ -12,6 +12,8 @@ requirejs.config({
         "datatables.bootstrap": "datatables-bootstrap3/BS3/assets/js/datatables",
         "datatables.scroller": "datatables-scroller/js/dataTables.scroller",
         "datatables.colReorder": "datatables-colreorder/js/dataTables.colReorder",
+        "jquery-mousewheel": "jquery-mousewheel/jquery.mousewheel",
```
Does it not work to use `jquery-mousewheel/jquery.mousewheel` and `datetimepicker/build/jquery.datetimepicker.full.min` as the dependencies in js? I forget what exactly it is about datatables extensions that necessitated this config, but most of our third-party code doesn't need to be added here.

biyeun replied:
it was complaining about this when I tried it (specifically with `jquery-mousewheel`), because somewhere else in datetimepicker it references `jquery-mousewheel` as the dependency...so it's definitely needed here. I figured if I was adding `jquery-mousewheel`, i might as well add `datetimepicker` since it was so long

orangejenny replied:
Cool, can we keep `mousewheel` here then but not the datetimepicker? I'd like "dependency names match filenames" to be as firm a rule as possible for simplicity's sake, even if it means typing out long names.

biyeun replied:
sure, done!

orangejenny replied:
thanks!

biyeun committed:
update to full path to datetimepicker

biyeun commented:
wooo dimagimon removed the risk labels on its own once master was merged in with the `datetimepicker_changes` ^_^
now this PR is still huge but officially low risk. yay!

millerdev reviewed (approved):
I think there are no blocking comments here, but the quality of this review doesn't feel very high because my eyes started to glaze over after a while.

Grumpy complaint because I'm exhausted: reviewing deleted code is frustrating in a huge PR like this. It would be kinder to reviewers to use `git rebase` to squash or fixup commits to eliminate code that is later deleted.

biyeun commented on corehq/apps/enterprise/decorators.py:
```diff
@@ -0,0 +1,12 @@
+def require_enterprise_admin(view_func):
+    @wraps(view_func)
+    def _inner(request, domain, *args, **kwargs):
+        # todo separate authorization from get method
```
no i was planning a different PR for this as that started to feel way out of scope

biyeun commented on corehq/apps/enterprise/decorators.py:
```diff
@@ -0,0 +1,12 @@
+    @wraps(view_func)
+    def _inner(request, domain, *args, **kwargs):
+        # todo separate authorization from get method
+        request.account = get_account_or_404(request, domain)
```
yes, that is my plan in a separate PR. I agree this name is very misleading based on how it's used and what originally made me panic when I was checking how the enterprise views were properly authenticated

biyeun commented on corehq/apps/sso/forms.py:
```diff
@@ -271,6 +272,14 @@ def __init__(self, identity_provider, *args, **kwargs):
         )
 
+        if self.idp.is_editable:
+            self.fields['is_editable'].help_text = mark_safe(
+                '<a href="{}">{}</a>'
+            ).format(
+                utils.get_dashboard_link(self.idp),
+                _("Edit Enterprise Settings")
+            )
```
it's definitely still used! I hear you on later deleted code bit, but I don't believe that's the case here. I think github collapsed the `forms.py` diff in the summary view because it was quite large.

millerdev replied:
It's not that `mark_safe` is not used, it's just not a best practice to use it that way.

biyeun replied:
sorry this was meant for the other comment. I'll update :)

biyeun commented on corehq/apps/sso/static/sso/js/edit_identity_provider.js:
```diff
@@ -33,11 +33,9 @@ hqDefine('sso/js/edit_identity_provider', [
-                    && newObject.indexOf('@') > 1
-                    && !newObject.endsWith('.');
+                // from http://stackoverflow.com/questions/46155/validate-email-address-in-javascript
+                var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/; // eslint-disable-line no-useless-escape
```
yes, I'm aware of this. I was thinking of creating a utility module in `hqwebapp`, but decided that was, again, another tangent and would leave it for cleanup.

millerdev replied:
:+1: on doing it as a cleanup step

biyeun commented on corehq/apps/sso/static/sso/js/enterprise_edit_identity_provider.js:
```diff
@@ -20,11 +20,9 @@ hqDefine('sso/js/enterprise_edit_identity_provider', [
-                    && newObject.indexOf('@') > 1
-                    && !newObject.endsWith('.');
+                // from http://stackoverflow.com/questions/46155/validate-email-address-in-javascript
+                var re = /^(([^<>()[\]\\.,;:\s@\"]+(\.[^<>()[\]\\.,;:\s@\"]+)*)|(\".+\"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/; // eslint-disable-line no-useless-escape
```
see comment above

millerdev replied:
I saw it. Four copies in our codebase seems excessive.

biyeun commented on corehq/apps/sso/utils.py:
```diff
@@ -23,3 +26,39 @@ def get_request_data(request):
+    return _get_full_sso_url("sso_saml_sls", identity_provider)
+
+
+def get_dashboard_link(identity_provider):
```
it's definitely still used! I hear you on later deleted code bit, but I don't believe that's the case here. I think github collapsed the forms.py diff in the summary view because it was quite large.

millerdev replied:
Yes, that explains it.

biyeun committed:
remove workaround for tests, use mock instead

biyeun committed:
use with instead of lambda for assertRaises in tests

biyeun commented:
what on earth happened. with that merge. I'm going to have to rebase. >_< but hopefully that's ok since daniel already reviewed and I addressed the feedback

biyeun commented on corehq/apps/sso/utils.py:
```diff
@@ -55,8 +55,11 @@ def get_dashboard_link(identity_provider):
+            args=(linked_subscription.subscriber.domain, identity_provider.slug,)
+        )
+    # would only ever reach this in tests
+    return '#'
```
done

biyeun replied:
done

biyeun commented on corehq/apps/enterprise/tests/test_decorators.py:
```diff
@@ -0,0 +1,66 @@
+        self.assertRaises(
+            Http404,
+            lambda: decorated_view(self.request, *self.request_args)
+        )
```
updated

biyeun committed:
use format_html instead of mark_safe

biyeun commented:
thank you @millerdev for all the quick feedback on this huge PR! all should be addressed and I'll wait for a final build ✅  before merging.

mjriley commented on corehq/apps/sso/models.py:
```diff
@@ -118,6 +118,23 @@ def renew_service_provider_certificate(self):
         self.date_sp_rollover_cert_expiration = None
         self.save()
 
+    @property
+    def email_domains(self):
+        return AuthenticatedEmailDomain.objects.filter(
+            identity_provider=self
+        ).values_list('email_domain', flat=True).all()
+
+    @property
+    def sso_exempt_users(self):
+        return UserExemptFromSingleSignOn.objects.filter(
+            email_domain__identity_provider=self,
+        ).values_list('username', flat=True)
```
What's the reasoning behind making these methods? The guideline may have changed, but to me, this breaks the principle of least surprise -- properties are lightweight things that are very efficient. If I'm making a database call, that is a function, and shouldn't be disguised as a property. I want to know that there might be some overhead to `get_email_domains()`, for example.

biyeun replied:
the reason is habit. I am referencing these properties in `jinja2` templates, and methods can't be called in templates like `{{ idp.get_email_domains() }}`, though it seems like now `jinja` will call it automatically if it's a function (eg `{{ idp.get_email_domains }}` works), but that wasn't always the case in the past.

biyeun replied:
that being said, I agree with your perspective and I'll update it in a followup PR

biyeun replied:
addressed in https://github.com/dimagi/commcare-hq/pull/29205

mjriley commented on corehq/apps/sso/utils.py:
```diff
@@ -23,3 +26,36 @@ def get_request_data(request):
+def get_dashboard_link(identity_provider):
+    from corehq.apps.accounting.models import Subscription
+    from corehq.apps.sso.views.enterprise_admin import EditIdentityProviderEnterpriseView
+    linked_subscription = Subscription.visible_objects.filter(
+        account=identity_provider.owner,
+        is_active=True,
+        account__is_active=True,
+    ).first()
```
use `get()`, not `first()`. `first()` is not semantically correct here. The reason I see some people prefer `first()` to `get()` is that they don't have to handle exceptions when the query returns nothing -- but that doesn't apply here, since the code is referencing into a possibly empty `linked_subscription`

biyeun replied:
well, another reason is that a customer account might have multiple active subscriptions, so `get()` wouldn't necessarily work here and would throw an error on those type of accounts. :)

mjriley replied:
Doh, you're 100% right. I wasn't reading that carefully enough, my bad.

mjriley commented on corehq/apps/sso/views/accounting_admin.py:
```diff
@@ -0,0 +1,226 @@
+        def _idp_to_row(idp):
+            edit_url = reverse(EditIdentityProviderAdminView.urlname, args=(idp.id,))
+            return [
+                f'<a href="{edit_url}">{idp.name}</a>',
```
This is an XSS vulnerability. The others were just good practice, this code is actually vulnerable to it.  Whenever you are working with HTML fragments, use `format_html()`

biyeun replied:
the idp name is always *only* available for accounting admins to edit, but yes I'll make sure to swap this to `format_html` in a followup PR

biyeun replied:
addressed in https://github.com/dimagi/commcare-hq/pull/29205

mjriley reviewed (commented):
Looks good. Love the thorough tests. I agree with Daniel, though, that the length of this PR made it difficult to focus on specific bits. There are definitely sections I had to skim over.  I think the most pressing comments are the `format_html()` ones, as I'm working on creating static analysis for this, and I'll be the one that has to go in and manually fix them later 😢

biyeun commented on corehq/apps/sso/views/accounting_admin.py:
```diff
@@ -0,0 +1,78 @@
+
+    @classmethod
+    def show_in_navigation(cls, domain=None, project=None, user=None):
+        return ENTERPRISE_SSO.enabled(user.username)
```
this it's part of the base class method's sig. some reports are used anonymously in places I guess? I'm not quite sure why `user=None`. Maybe it's for when only `domain` or `project` are available (which would never be the case for an accounting admin UI)

mjriley replied:
If `user=None` is really just a vestige of old code, and we're confident that the user can never actually be None, can we leave a comment indicating that it can't really be None? Otherwise I think it that case needs to be handled in some way.

biyeun replied:
it can be `None` though if `domain` or `project` are filled in. it depends on the report, and it's been so mangled I'm not 100% sure what the current state is and no there aren't enough tests to reliably make the change LOL 😭

biyeun commented on corehq/apps/sso/forms.py:
```diff
@@ -0,0 +1,89 @@
+            created_by=admin_user.username,
+            last_modified_by=admin_user.username,
+        )
+        idp.create_service_provider_certificate()
```
`create_service_provider_certificate` is re-used in quite a few other places, and I like to keep it separate as some tests don't need all that extra logic to be run. the sp cert is currently not being used and thus not currently required (though it is a best practice to create it). If it's not present on creation an automated task will catch it later. I am going to do a follow-up PR to allow enterprise admins to enable a "require full encryption" flag, which means updating the `IdentityProvider` model. This is due to the fact that not every Azure AD subscription level supports this. You need the "enterprise" subscription and I don't want to trial that out just yet.

biyeun commented on corehq/apps/sso/views/accounting_admin.py:
```diff
@@ -76,3 +85,49 @@ def _identity_providers(self):
+
+    @property
+    @memoized
+    def create_idp_form(self):
```
this is currently the pattern we have for other areas where we use forms in class-based views.

biyeun commented on corehq/apps/enterprise/decorators.py:
```diff
@@ -0,0 +1,12 @@
+    @wraps(view_func)
+    def _inner(request, domain, *args, **kwargs):
+        # todo separate authorization from get method
+        request.account = get_account_or_404(request, domain)
```
i did not name the original function. i am simply adding it to the decorator and will clean it up afterward. i agree it is poorly named. the authorization logic should happen in the decorator, not in a `get`

mjriley replied:
I think I understand -- when you say clean up, you mean that it will likely return a different HTTP code than 404?

biyeun replied:
I think `get_account_or_404` makes sense...`404` vs `403` is fine here...because it's not checking permissions but the existence of the account. However, the **permissions** check for whether the current user has enterprise admin privileges should absolutely not live there! Our pattern is to do permissions checks in decorators, and this clearly does not follow that pattern

biyeun replied:
see this PR for refactor. going to leave it open for review and trigger QA next week https://github.com/dimagi/commcare-hq/pull/29207

biyeun commented on corehq/apps/sso/async_handlers.py:
```diff
@@ -0,0 +1,179 @@
+            )
+        existing_email_domain.delete()
+
+    @property
+    @memoized
+    def idp_slug(self):
+        return self.data.get('requestContext[idpSlug]')
+
+    @property
+    @memoized
+    def identity_provider(self):
+        return IdentityProvider.objects.get(slug=self.idp_slug)
+
+    @property
+    @memoized
+    def email_domain(self):
+        return self.data.get('objectName')
```
easier to read, setup, and subclass imo. this is the pattern we've followed for class-based views

biyeun replied:
also, it wouldn't work to call it in `__init__` actually. see
https://stackoverflow.com/questions/46599145/django-class-view-init

The common recommendation is actually to NEVER start from `__init__` with class-based views.

mjriley replied:
Thanks for that link -- I've been looking for a reason to dig deeper into class-based views, and this let me get into the nitty-gritty. I'm guessing this pattern is a historical artifact, caused by having no logical place to initialize properties. The article you linked suggested overriding `dispatch()`, and that used to be the main way to do it. I'd imagine we went for memoized properties because it was slightly more intuitive than overriding dispatch. The correct way now, it seems, is to [override `setup()`](https://docs.djangoproject.com/en/3.1/ref/class-based-views/base/#django.views.generic.base.View.setup). However, that was only [introduced in 2.2](https://docs.djangoproject.com/en/3.1/releases/2.2/#generic-views). I get why we did it in the past, but I have a hard time advocating for solutions just because that's how it was done previously.

biyeun commented on corehq/apps/sso/tests/test_async_handlers.py:
```diff
@@ -0,0 +1,315 @@
+        with self.assertRaises(AsyncHandlerError):
+            handler.add_object()
+
+    def test_add_object_raises_error_if_conflict_with_another_idp(self):
```
`vwx.link` is an email domain for `idp_two` not `idp_one`

biyeun commented on corehq/apps/sso/tests/test_async_handlers.py:
```diff
@@ -0,0 +1,315 @@
+        with self.assertRaises(AsyncHandlerError):
+            handler.add_object()
+
+    def test_add_object_raises_errors_if_email_domain_is_linked_to_different_idp(self):
+        self.request.POST = {
+            'requestContext[idpSlug]': self.idp_one.slug,
+            'objectName': 'b@vwx.link',
+        }
+        handler = SSOExemptUsersAdminAsyncHandler(self.request)
+        with self.assertRaises(AsyncHandlerError):
+            handler.add_object()
```
I want the idp to be guaranteed fresh each time because I do make modifications to it (`is_active=True/False` etc) in certain tests.

biyeun replied:
I think I misread this. for this test the idp is created in `setUpClass`, and I read `setUp`. Generally, if I don't have to duplicate database calls I try to avoid it. I think @millerdev has said in the past `setUpClass` is always preferred if it's not necessary to have fresh data for each test.

mjriley replied:
Curious to hear @millerdev 's thoughts, but I'd consider that premature optimization.  We're essentially talking about trading readability for speed, right? I think that a test that contains all of its phases ([the AAA pattern](https://freecontent.manning.com/making-better-unit-tests-part-1-the-aaa-pattern/)) in its body is so much nicer to deal with than one where I need to hunt through setup or subclasses. While that link does mention extracting common elements into a constructor/setup method, it also mentions:

"The other drawback to extracting the arrangement code into the constructor is diminished test readability. You no longer see the full picture by looking at the test. You have to examine different places in the class to understand what the test method does.

Even if there’s not much arrangement logic, say, only instantiation of the fixtures, you’re still better off moving it directly to the test method. Otherwise, you’ll be wondering if it’s instantiation or there’s something else being configured there too. A self-contained test doesn’t leave you with such uncertainties."

biyeun replied:
yes, I do get where you're coming from. From that point of view, with tests it's better to be verbose rather than optimized with the goal of improved readability. However, I believe we take the over-optimized route because we also have a situation where our tests take forever to run and every millisecond adds up (limited Travis jobs). As a result, the tendency is to overly optimize test code and forgo some instant readability. Perhaps what's better is more explicit documentation/comments? also curious to hear Daniel's thoughts

biyeun commented on corehq/apps/sso/forms.py:
```diff
@@ -0,0 +1,550 @@
+
+    @transaction.atomic
+    def create_identity_provider(self, admin_user):
+        owner = BillingAccount.objects.get(id=self.cleaned_data['owner'])
```
i'm a little confused? `is_valid` should be called in the view...

mjriley replied:
Sorry, I basically mean defensive programming. With a method called `create_identity_provider()`, it's reasonable to assume all you need to do is create a form, supply data to the fields, and then call the method.  The implicit knowledge is that the user is expected to know that someone needs to have called `is_valid()` on that form first, because attempting to access these `self.cleaned_data` fields will raise an exception, otherwise. To prevent this, you could do something like the following:

```
def create_identity_provider(self, admin_user):
  if not self.is_valid():
    return None # or raise exception
  owner = BillingAccount.objects.get(...)
```

Now the class can't be misused. The downside is we're making an extra `is_valid()` call, but the documentation suggests this doesn't do anything if its already been run.

biyeun replied:
i get what you are saying. I'm just wondering if it's really needed considering the
```
if form.is_valid():
   # do something to process it
   return redirect("foo")
```

Is a pretty standard pattern when dealing with django forms and I don't think there would be any expectation to call `create_identity_provider` on a form without following that pattern?

biyeun commented on corehq/apps/sso/forms.py:
```diff
@@ -0,0 +1,550 @@
+        )
+        self.fields.update(sp_details_form.fields)
+
+        from corehq.apps.accounting.views import ManageBillingAccountView
```
yes, circular dependencies.

biyeun commented on corehq/apps/enterprise/tests/test_decorators.py:
```diff
@@ -0,0 +1,64 @@
+            name="test-enterprise-001",
+            is_active=True
+        )
+        cls.domain.save()
```
This is a pretty common pattern in quite a lot of our tests, so I don't think that's ever been a problem in practice if that is the case. I can definitely see the concern...but I vaguely remember we do some rudimentary cleanup of things like `Domains` in the test couch db before restarting tests. but maybe I'm wrong...? With that in mind, if it's been working for so many other tests, I don't see why it would suddenly fail here.

mjriley replied:
As an anecdotal counterexample, I've already fallen victim to this multiple times. I've had one change in behavior cause setUpClass() to fail, which then repeatedly failed afterwards, since the created object was never deleted. I also personally sometimes just terminate a debugging session, in which case, the tearDown method never runs.  I believe this was the main reason you see a lot of our generator functions generating random names -- since we can't guarantee  couch objects get deleted, we try to use unique identifiers which will never cause a collision. This comes with its own issues, however, and I think the cleanest solution is to just handle situations where your object already exists.

biyeun commented on corehq/apps/sso/forms.py:
```diff
@@ -0,0 +1,550 @@
+                )
+        return is_active
+
+    @transaction.atomic
+    def update_identity_provider(self, admin_user):
+        self.idp.name = self.cleaned_data['name']
+        self.idp.slug = self.cleaned_data['slug']
+        self.idp.is_editable = self.cleaned_data['is_editable']
+        self.idp.is_active = self.cleaned_data['is_active']
+        self.idp.last_modified_by = admin_user.username
+        self.idp.save()
+        return self.idp
```
it's out of habit & patterns that I decorate any `update/create` methods in forms with `@transaction.atomic`. What's the downside if it is included?

mjriley replied:
It's definitely slower, but unsure if its anything noticeable. On the SQL side, it's turning 1 `UPDATE` statement into 3 separate database calls:
```
BEGIN
UPDATE
COMMIT
```
Django might do something smart under the hood to batch the calls, but a naive implementation is going to issue 3 separate calls for those statements, so that's 3 round-trips vs 1.

[This answer](https://stackoverflow.com/questions/60166976/why-is-postgres-transaction-block-so-much-slower-and-how-to-solve/60170238#60170238) suggests that the additional log writing for small transactions significantly affects their cost, but I admit I don't know how well Postgres handles this, or the degree of the difference.

Other than that, it's mostly a conceptual cost. [The docs](https://docs.djangoproject.com/en/3.1/topics/db/transactions/#controlling-transactions-explicitly) have a bit near the bottom: "Open transactions have a performance cost for your database server. To minimize this overhead, keep your transactions as short as possible. This is especially important if you’re using atomic() in long-running processes, outside of Django’s request / response cycle."  I do see the argument that this can protect methods when future database calls are added, but I tend to treat transactions with caution, similar to locks and try/except blocks -- only the bare minimum of code lives in that block. That makes me wary of labelling non-transactional code as transactional.

biyeun commented:
@mjriley thanks for your thorough review! Always very helpful and much appreciated. I do still want to chat about the over-optimize issue with tests and the class-based view comments. For the more pressing feedback related to `format_html`, I've opened this PR and addressed a few other points
https://github.com/dimagi/commcare-hq/pull/29205

biyeun commented on corehq/apps/sso/forms.py:
```diff
@@ -0,0 +1,550 @@
+        if self.show_rollover_cert:
+            shown_fields.append(hqcrispy.B3TextField(
+                'sp_rollover_cert',
+                (f'<a href="?sp_rollover_cert_public" target="_blank">{download}</a>'
```
addressed in https://github.com/dimagi/commcare-hq/pull/29205

biyeun commented on corehq/apps/sso/forms.py:
```diff
@@ -0,0 +1,550 @@
+            shown_fields.extend([
+                hqcrispy.B3TextField(
+                    'sp_public_cert',
+                    f'<a href="?sp_cert_public" target="_blank">{download}</a>',
```
addressed in https://github.com/dimagi/commcare-hq/pull/29205

biyeun commented on corehq/apps/sso/forms.py:
```diff
@@ -0,0 +1,550 @@
+        help_link = "#"  # todo
+        help_text = _('<a href="{}">Please read this guide</a> on how to set up '
+                      'CommCare HQ with Azure AD.<br />You will need the following '
+                      'information:').format(help_link)
```
addressed in https://github.com/dimagi/commcare-hq/pull/29205

biyeun commented on corehq/apps/sso/forms.py:
```diff
@@ -0,0 +1,550 @@
+        self.idp = identity_provider
+        # todo eventually have a setting for IdentityProvider toggles based on
+        #  whether SP signing is enforced (dependent on client's Azure tier)
+        self.show_public_cert = kwargs.pop('show_public_cert', True)
+        self.show_rollover_cert = kwargs.pop('show_rollover_cert', True)
+        self.show_help_block = kwargs.pop('show_help_block', True)
```
addressed in https://github.com/dimagi/commcare-hq/pull/29205

biyeun commented on corehq/apps/sso/utils.py:
```diff
@@ -23,3 +26,22 @@ def get_request_data(request):
     }
+
+
+def _get_full_sso_url(view_name, identity_provider):
```
responded here https://github.com/dimagi/commcare-hq/pull/29217
